packages/firestore/src/util/bundle_reader.ts

```diff
@@ -0,0 +1,229 @@
+  }
+
+  /**
+   * Returns the next BundleElement (together with its byte size in the bundle)
+   * that has not been read from underlying ReadableStream. Returns null if we
+   * have reached the end of the stream.
+   *
+   * Throws an error if the first element is not a BundleMetadata.
+   */
+  async nextElement(): Promise<SizedBundleElement | null> {
+    const element = await this.readNextElement();
+    if (!element) {
+      return element;
+    }
+
+    if (!this.metadata) {
+      if (element.isBundleMetadata()) {
+        this.metadata = element.payload.metadata;
+      } else {
+        this.raiseError(
+          `The first element of the bundle is not a metadata, it is ${JSON.stringify(
+            element.payload
+          )}`
+        );
+      }
+    }
+
+    return element;
+  }
+
+  /**
+   * Reads from the head of internal buffer, and pulling more data from underlying stream if a complete element
+   * cannot be found, until an element(including the prefixed length and the JSON string) is found.
+   *
+   * Once a complete element is read, it is dropped from internal buffer.
+   *
+   * Returns either the bundled element, or null if we have reached the end of the stream.
+   */
+  private async readNextElement(): Promise<SizedBundleElement | null> {
+    const lengthBuffer = await this.readLength();
+    if (lengthBuffer === null) {
+      return null;
+    }
+
+    const lengthString = this.textDecoder.decode(lengthBuffer);
+    const length = Number(lengthString);
+    if (isNaN(length)) {
+      this.raiseError(`length string (${lengthString}) is not valid number`);
+    }
+
+    const jsonString = await this.readJsonString(lengthBuffer.length, length);
+    // Update the internal buffer to drop the read length and json string.
+    this.buffer = this.buffer.slice(lengthBuffer.length + length);
+
+    return new SizedBundleElement(
+      JSON.parse(jsonString),
+      lengthBuffer.length + length
+    );
+  }
+
+  // First index of '{' from the underlying buffer.
+  private indexOfOpenBracket(): number {
+    return this.buffer.findIndex(v => v === '{'.charCodeAt(0));
+  }
+
+  // Reads from the beginning of the internal buffer, until the first '{', and return
+  // the content.
+  // If reached end of the stream, returns a null.
+  private async readLength(): Promise<Uint8Array | null> {
+    let position = this.indexOfOpenBracket();
+    while (position < 0) {
+      const bytesRead = await this.pullMoreDataToBuffer();
+      if (bytesRead < 0) {
+        if (this.buffer.length === 0) {
+          return null;
+        }
+        position = this.indexOfOpenBracket();
+        // Underlying stream is closed, and we still cannot find a '{'.
+        if (position < 0) {
+          this.raiseError(
+            'Reached the end of bundle when a length string is expected.'
+          );
+        }
+      } else {
+        position = this.indexOfOpenBracket();
+      }
+    }
+
+    return this.buffer.slice(0, position);
```

Comment:
Should this method advance the buffer as well and return the length as a number, rather than returning the encoded byte array (this is not part of my change)?